### PR TITLE
Re-exports in ngtsc

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/program.ts [hunk 1]
```diff
@@ -272,9 +273,15 @@ export class NgtscProgram implements api.Program {
...
-    const beforeTransforms =
-        [ivyTransformFactory(compilation, this.reflector, this.importRewriter, this.isCore)];
-    const afterDeclarationsTransforms = [declarationTransformFactory(compilation)];
+    const beforeTransforms = [
+      ivyTransformFactory(compilation, this.reflector, this.importRewriter, this.isCore),
+      aliasTransformFactory(compilation.exportStatements) as ts.TransformerFactory<ts.SourceFile>,
+    ];
+    const afterDeclarationsTransforms = [
+      declarationTransformFactory(compilation),
+      aliasTransformFactory(compilation.exportStatements),
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/program.ts]: nit: may be add a comment on why we run `aliasTransformFactory` again as a part of `afterDeclarationsTransforms`?

File: packages/compiler-cli/src/ngtsc/scope/src/dependency.ts [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        } else {
+          // It's necessary to rewrite the `Reference`s to add alias expressions. This way, imports
+          // generated to these directives and pipes will use a shallow import to `sourceFile`
+          // instead of a deep import directly to the directive or pipe class.
+          //
+          // One important check here is whether the directive/pipe is declared in the same
+          // source file as the re-exporting NgModule. This can happen if both a directive, its
+          // NgModule, and the re-exporting NgModule are all in the same file. In this case,
+          // no import alias is needed as it would go to the same file anyway.
+          for (const directive of exportScope.exported.directives) {
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/scope/src/dependency.ts]: The logic to handle Directives and Pipes looks very similar. May be extract it out to a separate function?

File: packages/compiler-cli/src/ngtsc/scope/src/dependency.ts [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+      exported: {directives, pipes},
+    };
+  }
+
+  /**
+   * Read the metadata from a class that has already been compiled somehow (either it's in a .d.ts
+   * file, or in a .ts file with a handwritten definition).
+   *
+   * @param clazz the class of interest
+   * @param ngModuleImportedFrom module specifier of the import path to assume for all
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/scope/src/dependency.ts]: Params docs don't seem to correspond actual function signature.

File: packages/compiler-cli/src/ngtsc/scope/src/dependency.ts [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+   * declarations
+   * stemming from this module.
+   */
+  private readModuleMetadataFromClass(ref: Reference<ts.Declaration>): RawDependencyMetadata|null {
+    const clazz = ref.node;
+    const resolutionContext = clazz.getSourceFile().fileName;
+    // This operation is explicitly not memoized, as it depends on `ngModuleImportedFrom`.
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/scope/src/dependency.ts]: `ngModuleImportedFrom` doesn't seem to exist in the scope of this fn, old comment?

File: packages/compiler-cli/src/ngtsc/scope/src/dependency.ts [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    // TODO(alxhub): investigate caching of .d.ts module metadata.
+    const ngModuleDef = this.reflector.getMembersOfClass(clazz).find(
+        member => member.name === 'ngModuleDef' && member.isStatic);
+    if (ngModuleDef === undefined) {
+      return null;
+    } else if (
+        // Validate that the shape of the ngModuleDef type is correct.
+        ngModuleDef.type === null || !ts.isTypeReferenceNode(ngModuleDef.type) ||
+        ngModuleDef.type.typeArguments === undefined ||
+        ngModuleDef.type.typeArguments.length !== 4) {
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/scope/src/dependency.ts]: May be add a comment on what `4` means or declare a const?

File: packages/compiler-cli/src/ngtsc/scope/src/local.ts [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+            reexports.push({
+              fromModule: ref.alias.value.moduleName !,
+              symbolName: ref.alias.value.name !,
+              asAlias: exportName,
+            });
+          } else {
+            const expr = this.refEmitter.emit(ref.cloneWithNoIdentifiers(), sourceFile);
+            if (!(expr instanceof ExternalExpr) || expr.value.moduleName === null ||
+                expr.value.name === null) {
+              throw new Error('Expected ExternalExpr');
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/scope/src/local.ts]: May be add more info into the error, like `exportName`, so it's easier to debug if needed?

File: packages/compiler-cli/src/ngtsc/scope/src/util.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    return [];
+  }
+  return def.elementTypes.map(element => {
+    if (!ts.isTypeQueryNode(element)) {
+      throw new Error(`Expected TypeQueryNode`);
+    }
+    const type = element.exprName;
+    const {node, from} = reflectTypeEntityToDeclaration(type, checker);
+    if (!ts.isClassDeclaration(node)) {
+      throw new Error(`Expected ClassDeclaration`);
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/scope/src/util.ts]: I'd propose to add more info into the error message, like `ngModuleImportedFrom` and id/name of the def if this info is available.

File: packages/compiler-cli/src/ngtsc/scope/test/dependency_spec.ts [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+      checker, new TypeScriptReflectionHost(checker), aliasGenerator);
+
+  // Resolver for the refs object.
+  const get = (target: {}, name: string): Reference<ts.ClassDeclaration> => {
+    for (const sf of program.getSourceFiles()) {
+      const symbol = checker.getSymbolAtLocation(sf) !;
+      const exportedSymbol = symbol.exports !.get(name as ts.__String);
+      if (exportedSymbol !== undefined) {
+        const decl = exportedSymbol.valueDeclaration as ts.ClassDeclaration;
+        const specifier = /.*node_modules\/(\w+)\/index\.d\.ts$/.exec(sf.fileName) ![1];
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/scope/test/dependency_spec.ts]: nit: this RegExp seems identical with the one used in `testHost.fileNameToModuleName` fn, may be move it in a const?

File: packages/compiler-cli/src/ngtsc/scope/test/dependency_spec.ts [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+    // Explicitly verify that the directive has the correct owning module.
+    expect(scope.exported.directives[0].ref.ownedByModuleGuess).toBe('declaration');
+  });
+
+  it('should write correct aliases for deep dependencies', () => {
+    const {resolver, refs} = makeTestEnv(
+        {
+          'deep': `
+        export declare class DeepDir {
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/scope/test/dependency_spec.ts]: nit: formatting (content should be moved to the right).

**Review by AndrewKushnir — APPROVED** [review]
Great work @alxhub 👍 Just left a few nits. Thank you.

File: packages/compiler-cli/src/ngtsc/program.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/program.ts]: I actually removed it. Putting the re-exports in the .d.ts file is good behavior, but it breaks tsickle :-/.

File: packages/compiler-cli/src/ngtsc/scope/src/dependency.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/scope/src/dependency.ts]: Ahaha, I did this independently of this comment :D nice!

File: packages/compiler-cli/src/ngtsc/scope/src/dependency.ts (hunk 3 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/scope/src/dependency.ts]: Fixed.

File: packages/compiler-cli/src/ngtsc/scope/src/dependency.ts (hunk 4 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/scope/src/dependency.ts]: Yep, fixed.

File: packages/compiler-cli/src/ngtsc/scope/src/dependency.ts (hunk 5 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/scope/src/dependency.ts]: I think the comment covers it - a correctly shaped ngModuleDef type should have 4 type arguments.

File: packages/compiler-cli/src/ngtsc/scope/test/dependency_spec.ts (hunk 8 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/scope/test/dependency_spec.ts]: Done.

File: packages/compiler-cli/src/ngtsc/scope/test/dependency_spec.ts (hunk 9 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/scope/test/dependency_spec.ts]: 👍

File: packages/compiler-cli/src/ngtsc/scope/src/local.ts (hunk 6 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/scope/src/local.ts]: Done.

File: packages/compiler-cli/src/ngtsc/scope/src/util.ts (hunk 7 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/scope/src/util.ts]: Done.

**Comment by alxhub** [comment]
Caretaker: I am the approver for compiler/ngcc. This PR is otherwise approved by @AndrewKushnir :)

**Pull request closed without merging**